### PR TITLE
[Doc][KubeRay] Kueue / RayJob - tutorial with toy example that everyone can try it locally

### PR DESCRIPTION
**Pull request by kevin85421**
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by architkulkarni — APPROVED** [review]
This is awesome and I love the concrete example with preemption.

I think it would be really great if we could link to this doc from somewhere in the ordinary Ray Jobs documentation, because job queuing is a very common user question. You can decide where to put it, but let me know if you need some guidance.

I'll defer to the Ray doc team for style edits.

**Comment by kevin85421** [comment]
@architkulkarni I will add a link to the "Next Steps" section in #43590 after this PR is merged.

**Comment by kevin85421** [comment]
cc @andrewsykim for review.

**Comment by architkulkarni** [comment]
Ah I was referring to the non-kuberay Ray job submission docs actually, but it'll be great to have a link in the kuberay doc as well!

**Comment by kevin85421** [comment]
cc @Yicheng-Lu-llll would you mind taking a look at this PR?

**Review by andrewsykim — APPROVED** [review]
LGTM -- I like the simple examples used. Thanks @kevin85421

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:4]:
```suggestion
This guide demonstrates the integration of KubeRay and Kueue for gang and priority scheduling using RayJob on a local Kind cluster.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:7]:
```suggestion
## Kueue
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:10]:
```suggestion
* To make a job wait.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:11]:
```suggestion
* To admit a job to start, which triggers Kubernetes to create pods.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:12]:
```suggestion
* To preempt a job, which triggers Kubernetes to delete active pods.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:35]:
```suggestion
Some limitations exist between Kueue and RayJob. See the [limitations of Kueue](https://kueue.sigs.k8s.io/docs/tasks/run_rayjobs/#c-limitations) for more details.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:95]:
```suggestion
  * The LocalQueue `user-queue` is a namespaced object in the `default` namespace which belongs to a ClusterQueue. A typical practice is to assign a namespace to a tenant, team, or user of an organization. Users submit jobs to a LocalQueue, instead of to a ClusterQueue directly.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:97]:
```suggestion
  * The WorkloadPriorityClass `prod-priority` has a higher value than the WorkloadPriorityClass `dev-priority`. RayJob custom resources with the `prod-priority` priority class take precedence over RayJob custom resources with the `dev-priority` priority class.  
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:107]:
```suggestion
Kueue admits workloads on an “all or nothing” basis, ensuring that Kubernetes never partially provisions a RayJob or RayCluster.
```

Comment (angelinalg) [reply doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:107]: Not sure if Kubernetes is the correct noun, so please correct as needed. Trying to remove passive voice.

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:108]:
```suggestion
Use gang scheduling strategy to avoid wasting compute resources caused by inefficient scheduling of workloads.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:116]:
```suggestion
Before creating the RayJob, modify the RayJob metadata with the following:
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:127]:
```suggestion
Note these important points for RayJob custom resources:
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:129]:
```suggestion
* The RayJob runs a simple Python script that demonstrates a loop running 600 iterations, printing the iteration number and sleeping for 1 second per iteration. Hence, the RayJob runs for about 600 seconds after the submitted Kubernetes Job starts.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:130]:
```suggestion
* Set `shutdownAfterJobFinishes` to true for RayJob to enable automatic cleanup. This setting triggers KubeRay to deletes the RayCluster after the RayJob finishes.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:140]:
```suggestion
Therefore, the second RayJob custom resource remains pending, and Kubernetes doesn't create Pods from the pending RayJob, even though the remaining resources are sufficient for a Pod.
```

Comment (angelinalg) [reply doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:140]: Again, not sure if the subject is Kubernetes, please correct as needed.

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:198]:
```suggestion
If you followed the previous step, the RayJob YAML manifest `ray-job.kueue-toy-sample.yaml` should already be set to the `dev-priority` priority class.
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:205]:
```suggestion
Before creating the RayJob with the higher priority class `prod-priority`, modify the RayJob metadata with the following:
```

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (angelinalg) [inline doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:221]:
```suggestion
You can see that KubeRay operator deletes the Pods belonging to the RayJob with the lower priority class `dev-priority` and creates the Pods belonging to the RayJob with the higher priority class `prod-priority`.
```

**Review by angelinalg — APPROVED** [review]
Just some nits. Sorry for the delay.

File: doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md
Comment (kevin85421) [reply doc/source/cluster/kubernetes/k8s-ecosystem/kueue.md:140]: Update "Kubernetes" to "KubeRay".

**Review by Yicheng-Lu-llll — APPROVED** [review]

**Comment by kevin85421** [comment]
The lint error is related to RLLib and has no relationship with this PR.
Screen Shot 2024-03-04 at 9 24 37 AM

**Pull request merged**
